### PR TITLE
Adds optional combinator, changes pipeline to required

### LINKED ISSUE #23: Optional columns
In our application there are some required columns the user submit and some are optional.

At the moment I encode this like this:

```elm
optional :Decoder a -> Decoder (Maybe a -> b) -> Decoder b
optional default childDecoder =
    pipeline (Csv.oneOf (Csv.map Just childDecoder) [ Csv.succeed Nothing ])

decoder =
    Csv.into Tuple.pair 
         |> pipeline (Csv.field "requiredColumn" Csv.float)
         |> optional (Csv.field "optionalColumn" Csv.float)

```

This sort of works, but has some subtle behavioural gotchas:

requiredColumn | optionalColumn
-----------------|----------------
1  | 2
2 | four
3 | 6

would successfully decode into:

```elm
[ ( 1, Just 2 ), ( 2, Nothing ), (3, Just 6) ]
```

which is not what I want. In this case I'd like to flag the error to the user, but I'd still like to mark some columns as optional. I can't really see a way to achieve this with the current API. So I'd propose something like 

```elm
optionalField : String -> Decoder a -> Decoder (Maybe a)
```

which would only go to the `Nothing` case iff the column was not present at all in the CSV.

### PR DESCRIPTION
**Pull request by gampleman**
Fixes #23.

### CODE REVIEW

File: src/Csv/Decode.elm [hunk 1]
```diff
@@ -761,11 +761,59 @@ into =
...
+required =
     map2 (\value fn -> fn value)
 
 
+{-| Like `required`, but tolerates if a column isn't provided at all.
+
+type alias Pet =
+{ id : Int
+, name : Maybe String
+}
```
Comment (BrianHicks) [inline src/Csv/Decode.elm:774]: I think the formatting got squished a little here

File: tests/Csv/DecodeTest.elm [hunk 2]
```diff
@@ -596,3 +596,18 @@ fromMaybeTest =
                             )
                         )
         ]
+
+
+optionalTest : Test
+optionalTest =
+    describe "optional"
```
Comment (BrianHicks) [inline tests/Csv/DecodeTest.elm:603]: could you add a test showing a successful decode when the field is present?

File: tests/Csv/DecodeTest.elm [hunk 3]
```diff
@@ -596,3 +596,18 @@ fromMaybeTest =
...
+        [ test "succeds when missing column" <|
```
Comment (BrianHicks) [inline tests/Csv/DecodeTest.elm:604]:
```suggestion
        [ test "succeeds when missing column" <|
```

just a tiny change here!

**Review by BrianHicks — CHANGES_REQUESTED** [review]
really great start. Thank you for doing this! Just a few things to fix and we'll be good to go.

**Comment by BrianHicks** [comment]
if we're making a major change, this one seems worth including to me. Are you interested in getting the conflicts fixed and getting it up and running again?

**Comment by gampleman** [comment]
Sorry, started looking into this, but didn't we already do https://github.com/BrianHicks/elm-csv/pull/30? This is sort of the same thing?

**Comment by BrianHicks** [comment]
Yeah, that’s fair. Thanks for pointing that out. I was just looking over open PRs and liked the optional/required distinction here. 

Ok to close then?

**Comment by gampleman** [comment]
Yeah let’s close

**Pull request closed without merging**